accounts/abi/type.go

```diff
@@ -40,53 +38,59 @@ const (
 
 // Type is the reflection of the supported argument type
 type Type struct {
-	IsSlice   bool
-	SliceSize int
+	IsSlice, IsArray bool
+	SliceSize        int
+
+	Elem *Type
+
+	Kind reflect.Kind
+	Type reflect.Type
+	Size int
+	T    byte // Our own type checking
 
-	Kind       reflect.Kind
-	Type       reflect.Type
-	Size       int
-	T          byte   // Our own type checking
 	stringKind string // holds the unparsed string for deriving signatures
 }
 
 var (
+	// fullTypeRegex parses the abi types
+	//
+	// Types can be in the format of:
+	//
+	// 	Input  = Type [ "[" [ Number ] "]" ] Name .
+	// 	Type   = [ "u" ] "int" [ Number ] .
+	//
+	// Examples:
+	//
+	//      string     int       uint       real
+	//      string32   int8      uint8      uint[]
+	//      address    int256    uint256    real[2]
 	fullTypeRegex = regexp.MustCompile("([a-zA-Z0-9]+)(\\[([0-9]*)?\\])?")
-	typeRegex     = regexp.MustCompile("([a-zA-Z]+)([0-9]*)?")
+	// typeRegex parses the abi sub types
+	typeRegex = regexp.MustCompile("([a-zA-Z]+)([0-9]*)?")
 )
 
-// NewType returns a fully parsed Type given by the input string or an error if it  can't be parsed.
-//
-// Strings can be in the format of:
-//
-// 	Input  = Type [ "[" [ Number ] "]" ] Name .
-// 	Type   = [ "u" ] "int" [ Number ] .
-//
-// Examples:
-//
-//      string     int       uint       real
-//      string32   int8      uint8      uint[]
-//      address    int256    uint256    real[2]
+// NewType creates a new reflection type of abi type given in t.
 func NewType(t string) (typ Type, err error) {
-	// 1. full string 2. type 3. (opt.) is slice 4. (opt.) size
-	// parse the full representation of the abi-type definition; including:
-	// * full string
-	// * type
-	// 	* is slice
-	//	* slice size
 	res := fullTypeRegex.FindAllStringSubmatch(t, -1)[0]
-
 	// check if type is slice and parse type.
 	switch {
 	case res[3] != "":
 		// err is ignored. Already checked for number through the regexp
 		typ.SliceSize, _ = strconv.Atoi(res[3])
-		typ.IsSlice = true
+		typ.IsArray = true
 	case res[2] != "":
 		typ.IsSlice, typ.SliceSize = true, -1
 	case res[0] == "":
 		return Type{}, fmt.Errorf("abi: type parse error: %s", t)
 	}
+	if typ.IsArray || typ.IsSlice {
+		sliceType, err := NewType(res[1])
+		if err != nil {
+			return Type{}, err
+		}
+		typ.Elem = &sliceType
+		return typ, nil
```

Comment:
It cannot be moved up easily because of the canonicalisation of uint.
A working fix would be:

```diff
diff --git a/accounts/abi/type.go b/accounts/abi/type.go
index cad7b82..ec466e1 100644
--- a/accounts/abi/type.go
+++ b/accounts/abi/type.go
@@ -89,6 +89,7 @@ func NewType(t string) (typ Type, err error) {
 			return Type{}, err
 		}
 		typ.Elem = &sliceType
+		typ.stringKind = sliceType.stringKind + t[len(res[1]):]
 		return typ, nil
 	}
 
@@ -110,6 +111,7 @@ func NewType(t string) (typ Type, err error) {
 		varSize = 256
 		t += "256"
 	}
+	typ.stringKind = t
 
 	switch varType {
 	case "int":
@@ -149,9 +151,7 @@ func NewType(t string) (typ Type, err error) {
 	default:
 		return Type{}, fmt.Errorf("unsupported arg type: %s", t)
 	}
-	typ.stringKind = t
-
-	return
+	return typ, nil
 }
 
 // String implements Stringer
```